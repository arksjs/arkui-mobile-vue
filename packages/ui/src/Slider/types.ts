import type { FormItemCommonProps } from '../Form/types'

type OnChange = (value: number) => void

export interface SlideCommonProps extends FormItemCommonProps {
  showValue?: boolean
  color?: string
  min?: number | string
  max?: number | string
  step?: number | string
}

export interface SliderProps extends SlideCommonProps {
  value?: number
}

export interface SliderEmits {
  onInput?: OnChange
  onChange?: OnChange
}
