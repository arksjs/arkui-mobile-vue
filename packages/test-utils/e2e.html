<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
