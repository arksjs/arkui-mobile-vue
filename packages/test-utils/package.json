{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
