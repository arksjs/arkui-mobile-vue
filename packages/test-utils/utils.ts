export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
