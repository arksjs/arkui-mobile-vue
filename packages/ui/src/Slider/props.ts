import { isNumeric } from '../helpers/util'
import { colorValidator } from '../helpers/validator'

export const slideProps = {
  min: {
    type: [Number, String],
    validator: isNumeric,
    default: 0
  },
  max: {
    type: [Number, String],
    validator: isNumeric,
    default: 100
  },
  step: {
    type: [Number, String],
    validator: isNumeric,
    default: 1
  },
  showValue: {
    type: Boolean,
    default: false
  },
  color: {
    type: String,
    validator: colorValidator
  }
}
