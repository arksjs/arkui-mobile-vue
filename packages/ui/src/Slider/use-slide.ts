import { computed } from 'vue'
import {
  getNumber,
  rangeNumber,
  addClassName,
  getRelativeOffset,
  removeClassName
} from '../helpers'
import { useTouch } from '../hooks'
import type { SlideCommonProps } from './types'
import { getSlideClasses, getSlideStyles } from './util'

interface Coords {
  prevValue: number // 滑动之前value值
  thumb: boolean // 是否点击了滑块
  thumbW: number // 滑块宽度
  clientStartX: number // 按下初始位置
  thumbXInTrack: number // 按下滑块在滑轨内的初始位置
  trackX: number // 按下滑轨的初始位置
  trackW: number // 滑轨宽度
  moved: boolean // 是否滑动
  $target: HTMLElement // 滑动元素
}

interface UseOptions {
  getValue: ($target: HTMLElement) => number
  move: (args: {
    value: number
    progress: number
    $target: HTMLElement
  }) => void
  end: (args: {
    value: number
    isChange: boolean
    $target: HTMLElement
  }) => void
}

const thumbW = 24

export function useSlide(
  props: SlideCommonProps,
  { move, end, getValue }: UseOptions
) {
  let coords: Coords | null
  let tempValue = 0

  function toInteger(number: number | string) {
    return Math.round(number as number)
  }

  function getMinMax() {
    let min = toInteger(getNumber(props.min, 0))
    let max = toInteger(getNumber(props.max, 100))

    if (min > max) {
      // 兼容min max搞反的问题
      max = [min, (min = max)][0]
    }

    return [min, max]
  }

  function value2Progress(val: number) {
    const [min, max] = getMinMax()

    return (val - min) / (max - min)
  }

  function rangeValue(val: number) {
    const [min, max] = getMinMax()

    return rangeNumber(val, min, max)
  }

  function updateByX(x: number, { trackW, prevValue, $target }: Coords) {
    x = rangeNumber(x, 0, trackW)

    const [min, max] = getMinMax()
    const step = toInteger(getNumber(props.step, 1))

    let newVal = toInteger(((max - min) * x) / trackW + min)
    newVal = toInteger((newVal - prevValue) / step) * step + prevValue
    newVal = rangeNumber(newVal, min, max)

    tempValue = newVal

    move({ value: tempValue, progress: value2Progress(newVal), $target })
  }

  const { onTouchStart, onTouchMove, onTouchEnd, onDragStart } = useTouch({
    onStart(e) {
      if (props.disabled) {
        // 禁用
        return
      }

      if (!e.touchCurrentElement) {
        return
      }

      const { clientX } = e.touchObject

      const $target = e.target as HTMLElement
      const trackRects = e.touchCurrentElement.getClientRects()[0]
      const thumb = !!$target.dataset.thumb

      coords = {
        prevValue: getValue($target),
        thumb,
        thumbW,
        clientStartX: clientX,
        thumbXInTrack: getRelativeOffset($target, e.touchCurrentElement)
          .offsetLeft,
        trackX: trackRects.left,
        trackW: trackRects.width - thumbW,
        moved: false,
        $target
      }

      if (thumb) {
        addClassName($target, 'active')
      }

      tempValue = coords.prevValue

      e.preventDefault()
    },
    onMove(e) {
      if (!coords) {
        return
      }

      coords.moved = true

      if (!coords.thumb) {
        return
      }

      const { clientX } = e.touchObject
      const { clientStartX, thumbXInTrack } = coords

      updateByX(thumbXInTrack + clientX - clientStartX, coords)

      e.stopPropagation()
    },
    onEnd(e) {
      if (coords) {
        if (!coords.thumb && !coords.moved) {
          updateByX(
            coords.clientStartX - coords.trackX - coords.thumbW / 2,
            coords
          )
        }

        if (coords.thumb) {
          removeClassName(coords.$target, 'active')
        }

        end({
          value: tempValue,
          isChange: coords.prevValue !== tempValue,
          $target: coords.$target
        })

        coords = null
        e.stopPropagation()
      }
    }
  })

  const slideClasses = computed(() => getSlideClasses(props.disabled))
  const slideStyles = computed(() => getSlideStyles(props.color))

  return {
    toInteger,
    rangeValue,
    value2Progress,
    getMinMax,
    slideClasses,
    slideStyles,

    onTouchStart,
    onTouchMove,
    onTouchEnd,
    onDragStart
  }
}
