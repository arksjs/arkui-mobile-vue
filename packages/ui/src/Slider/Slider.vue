<template>
  <div :class="slideClasses" :style="slideStyles">
    <div class="fx-slider_inner" ref="sliderEl">
      <div class="fx-slider_box">
        <div
          class="fx-slider_track"
          :style="{ width: progress * 100 + '%' }"
        ></div>
        <div
          class="fx-slider_thumb"
          data-thumb="true"
          :style="{ left: progress * 100 + '%' }"
        >
          {{ showValue ? inputValue : '' }}
        </div>
      </div>
      <input
        type="hidden"
        :name="name"
        :disabled="disabled"
        :value="inputValue"
        ref="input"
      />
    </div>
  </div>
</template>

<script lang="ts">
import { ref, defineComponent, watch, nextTick } from 'vue'
import { isNumeric } from '../helpers/util'
import { formNumberValueEmits, formItemProps } from '../Form/form'
import { slideProps } from '../Slider/props'
import { useSlide } from '../Slider/use-slide'

export default defineComponent({
  name: 'fx-slider',
  props: {
    ...formItemProps,
    ...slideProps,
    modelValue: {
      type: [Number, String],
      validator: isNumeric
    }
  },
  emits: {
    ...formNumberValueEmits
  },
  setup(props, ctx) {
    const progress = ref(0)
    const inputValue = ref(0)
    const { emit } = ctx

    const {
      sliderEl,
      toInteger,
      rangeValue,
      value2Progress,
      slideClasses,
      slideStyles
    } = useSlide(props, {
      getValue() {
        return inputValue.value
      },
      move({ value: newVal, progress: newProgress }) {
        inputValue.value = newVal
        progress.value = newProgress

        emitModel()
        emit('input', inputValue.value)
      },
      end({ isChange }) {
        isChange && emit('change', inputValue.value)
      }
    })

    function emitModel() {
      if (
        props.modelValue == null ||
        inputValue.value !== toInteger(props.modelValue)
      ) {
        emit('update:modelValue', inputValue.value)
        return true
      }
      return false
    }

    function updateValue(val: unknown) {
      if (val == null) {
        return
      }

      let newVal = toInteger(val as string)

      if (isNaN(newVal)) {
        return
      }

      newVal = rangeValue(newVal)

      if (newVal !== inputValue.value) {
        inputValue.value = newVal
        progress.value = value2Progress(newVal)
      }
    }

    watch(
      () => props.modelValue,
      val => updateValue(val)
    )

    watch([() => props.min, () => props.max], () => {
      nextTick(() => {
        updateValue(inputValue.value)

        if (emitModel()) {
          emit('change', inputValue.value)
        }
      })
    })

    updateValue(props.modelValue || 0)

    if (emitModel()) {
      emit('change', inputValue.value)
    }

    return {
      sliderEl,
      progress,
      inputValue,
      slideClasses,
      slideStyles
    }
  }
})
</script>
