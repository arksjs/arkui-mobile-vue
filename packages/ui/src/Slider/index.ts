import { withInstall } from '../helpers/with-install'
import Slider from './Slider.vue'

export { Slider }
export const install = withInstall(Slider)
export default Slider
