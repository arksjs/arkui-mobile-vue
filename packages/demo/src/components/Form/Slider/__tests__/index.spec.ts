import { shallowMount } from '@vue/test-utils'
import { Slider } from '@/Slider'

describe('Slider', () => {
  test('should render default correctly', () => {
    const wrapper = shallowMount(Slider, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop showValue=true correctly', () => {
    const wrapper = shallowMount(Slider, {
      props: {
        showValue: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop disabled=true correctly', () => {
    const wrapper = shallowMount(Slider, {
      props: {
        disabled: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
