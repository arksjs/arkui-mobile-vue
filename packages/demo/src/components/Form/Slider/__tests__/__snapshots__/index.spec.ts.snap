// Vitest Snapshot v1

exports[`Slider > should render default correctly 1`] = `
"<div class=\\"fx-slider\\">
  <div class=\\"fx-slider_inner\\">
    <div class=\\"fx-slider_box\\">
      <div class=\\"fx-slider_track\\" style=\\"width: 0%;\\"></div>
      <div class=\\"fx-slider_thumb\\" data-thumb=\\"true\\" style=\\"left: 0%;\\"></div>
    </div><input type=\\"hidden\\" name=\\"\\" value=\\"0\\">
  </div>
</div>"
`;

exports[`Slider > should render prop disabled=true correctly 1`] = `
"<div class=\\"fx-slider disabled\\">
  <div class=\\"fx-slider_inner\\">
    <div class=\\"fx-slider_box\\">
      <div class=\\"fx-slider_track\\" style=\\"width: 0%;\\"></div>
      <div class=\\"fx-slider_thumb\\" data-thumb=\\"true\\" style=\\"left: 0%;\\"></div>
    </div><input type=\\"hidden\\" name=\\"\\" disabled=\\"\\" value=\\"0\\">
  </div>
</div>"
`;

exports[`Slider > should render prop showValue=true correctly 1`] = `
"<div class=\\"fx-slider\\">
  <div class=\\"fx-slider_inner\\">
    <div class=\\"fx-slider_box\\">
      <div class=\\"fx-slider_track\\" style=\\"width: 0%;\\"></div>
      <div class=\\"fx-slider_thumb\\" data-thumb=\\"true\\" style=\\"left: 0%;\\">0</div>
    </div><input type=\\"hidden\\" name=\\"\\" value=\\"0\\">
  </div>
</div>"
`;
